Wstępne przetwarzanie danych 

In [2]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv("reviews_all.csv")

In [3]:
df1 = pd.read_excel("review_links_books_1a.xlsx")

In [4]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5972 entries, 0 to 5971
Data columns (total 1 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Link do recenzji  5972 non-null   object
dtypes: object(1)
memory usage: 46.8+ KB


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5974 entries, 0 to 5973
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Unnamed: 0          5974 non-null   int64  
 1   Link                5974 non-null   object 
 2   Autor               5970 non-null   object 
 3   Tytuł               5970 non-null   object 
 4   Data_recenzji       5964 non-null   object 
 5   Treść_recenzji      5974 non-null   object 
 6   Średnia_ocen        5970 non-null   float64
 7   Liczba_ocen         5969 non-null   float64
 8   Liczba_czytelników  5970 non-null   float64
dtypes: float64(3), int64(1), object(5)
memory usage: 420.2+ KB


In [ ]:
df.columns

Index(['Unnamed: 0', 'Link', 'Autor', 'Tytuł', 'Data_recenzji',
       'Treść_recenzji', 'Średnia_ocen', 'Liczba_ocen', 'Liczba_czytelników'],
      dtype='object')

# W pliku wyjściowym z pobranymi linkami do recenzji mam 5972 rekordy.
# W pliku "ostatecznym" mam 2 rekordy więcej.
# Prawdopodobnie podczas pobierania recenzji, któreś linki się zdublowały.
# Usunę duble (może jest ich więcej)

In [ ]:
# Usywam duplikaty
df = df.drop_duplicates(["Link", "Autor", "Tytuł", "Data_recenzji", "Treść_recenzji", "Średnia_ocen", "Liczba_ocen", "Liczba_czytelników"])

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5972 entries, 0 to 5973
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Unnamed: 0          5972 non-null   int64  
 1   Link                5972 non-null   object 
 2   Autor               5968 non-null   object 
 3   Tytuł               5968 non-null   object 
 4   Data_recenzji       5962 non-null   object 
 5   Treść_recenzji      5972 non-null   object 
 6   Średnia_ocen        5968 non-null   float64
 7   Liczba_ocen         5967 non-null   float64
 8   Liczba_czytelników  5968 non-null   float64
dtypes: float64(3), int64(1), object(5)
memory usage: 466.6+ KB


# Na chwilę obecną mam 5972 rekordy.
# Teraz typ zmiennych 'Liczba_ocen', 'Liczba_czytelników' zamienię na int.
# Typ zmiennej 'Data_recenzji' zmienię na date.

In [ ]:
# Zmieniam typ zmiennej "Liczba_ocen"
df["Liczba_ocen"] = df["Liczba_ocen"].fillna(0).astype(int)

In [ ]:
# Zmieniam typ zmiennej "Liczba_czytelników"
df["Liczba_czytelników"] = df["Liczba_czytelników"].fillna(0).astype(int)

In [ ]:
# Zmieniam typ zmiennej "Data_recenzji"
df["Data_recenzji"] = pd.to_datetime(df["Data_recenzji"], format="%Y-%m-%d")

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5972 entries, 0 to 5973
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Unnamed: 0          5972 non-null   int64         
 1   Link                5972 non-null   object        
 2   Autor               5968 non-null   object        
 3   Tytuł               5968 non-null   object        
 4   Data_recenzji       5962 non-null   datetime64[ns]
 5   Treść_recenzji      5972 non-null   object        
 6   Średnia_ocen        5968 non-null   float64       
 7   Liczba_ocen         5972 non-null   int32         
 8   Liczba_czytelników  5972 non-null   int32         
dtypes: datetime64[ns](1), float64(1), int32(2), int64(1), object(4)
memory usage: 419.9+ KB


In [ ]:
# Resetuję indeksy
df = df.reset_index(drop=True)

In [ ]:
# Usuwam pierwsza kolumnę "Unnamed: 0"
df = df.drop(df.columns[0], axis=1)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5972 entries, 0 to 5971
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Link                5972 non-null   object        
 1   Autor               5968 non-null   object        
 2   Tytuł               5968 non-null   object        
 3   Data_recenzji       5962 non-null   datetime64[ns]
 4   Treść_recenzji      5972 non-null   object        
 5   Średnia_ocen        5968 non-null   float64       
 6   Liczba_ocen         5972 non-null   int32         
 7   Liczba_czytelników  5972 non-null   int32         
dtypes: datetime64[ns](1), float64(1), int32(2), object(4)
memory usage: 326.7+ KB


# W kilku przypadkach mam braki danych

In [ ]:
# Wyświetlam braki danych we wszystkich zmiennych
df[df.isna().any(axis=1)]

Link                 Autor  \
3142  https://lubimyczytac.pl/oficjalne-recenzje-ksi...       Brigid Kemmerer   
3190  https://lubimyczytac.pl/oficjalne-recenzje-ksi...  Agnieszka Stelmaszyk   
4365  https://lubimyczytac.pl/oficjalne-recenzje-ksi...        Colleen Hoover   
4507  https://lubimyczytac.pl/oficjalne-recenzje-ksi...         Lewis Carroll   
4700  https://lubimyczytac.pl//6048/dawno-temu-na-li...                   NaN   
4751          https://lubimyczytac.pl//5918/syreni-czar                   NaN   
4766  https://lubimyczytac.pl//5870/milosc-w-cieniu-...                   NaN   
4767  https://lubimyczytac.pl//5865/tam-kolo-mlyna-r...                   NaN   
4932  https://lubimyczytac.pl/oficjalne-recenzje-ksi...           Philip Roth   
5420  https://lubimyczytac.pl/oficjalne-recenzje-ksi...             Shaun Tan   

                                                  Tytuł Data_recenzji  \
3142                                 Listy do utraconej           NaT   
3190              Imaginarium. Tom 1. Gildia Wynalazców           NaT   
4365                                          Ugly Love           NaT   
4507  Po drugiej stronie Lustra i co tam Alicja znal...           NaT   
4700                                                NaN           NaT   
4751                                                NaN           NaT   
4766                                                NaN           NaT   
4767                                                NaN           NaT   
4932                               Amerykańska sielanka           NaT   
5420                               Zgubione, znalezione           NaT   

                                         Treść_recenzji  Średnia_ocen  \
3142  Juliet, zostawia odręcznie pisane listy na gro...           8.2   
3190  Zwariowana i niepokorna Marianna w zielonym ko...           8.4   
4365  Miłość została wpleciona do tysięcy fabuł ksią...           7.9   
4507  „Alicja” Lewisa Carrolla jest tak bardzo obecn...           7.1   
4700                                                              NaN   
4751                                                              NaN   
4766                                                              NaN   
4767                                                              NaN   
4932  Pewna pokaźna część ludzkości od przeszło już ...           7.4   
5420  Shaun Tan jest kapłanem wyobcowania. To znaczy...           8.2   

      Liczba_ocen  Liczba_czytelników  
3142         1554                3577  
3190           65                 182  
4365        10893               20529  
4507         1205                2442  
4700            0                   0  
4751            0                   0  
4766            0                   0  
4767            0                   0  
4932          674                2338  
5420           91                 221

**NaN (Not a Number); NaT (Not a Time)

# W 4 przypadkach nie mam pobranych żadnych recenzji. Sprawdziłam, i wiem, że te recenzje zostały usunięte. Więc te 4 rekordy usunę (index 4700, 4751, 4766, 4767)

In [ ]:
# Usuwam rekordy dla indeksów: 4700, 4751, 4766, 4767
to_drop = [4700, 4751, 4766, 4767]
df = df.drop(to_drop, axis=0)

# Teraz uzupełnię braki w datach recenzji

In [ ]:
# Uzupełniam braki danych w datach na podstawie najbliższych rekordów.
# Użyję metody "ffill" - uzupełnia braki danych wartościami z poprzednich rekordów (metoda "bfill" z następnych rekordów).
df["Data_recenzji"] = df["Data_recenzji"].fillna(method="ffill")

C:\Users\konta\AppData\Local\Temp\ipykernel_11596\2618223451.py:3: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df['Data_recenzji'] = df['Data_recenzji'].fillna(method='ffill')


In [ ]:
df.iloc[5415]

Link                  https://lubimyczytac.pl/oficjalne-recenzje-ksi...
Autor                                                     Sue Monk Kidd
Tytuł                                                   Czarne skrzydła
Data_recenzji                                       2014-09-11 00:00:00
Treść_recenzji        Pamiętam, że serial „Północ-Południe” (na pods...
Średnia_ocen                                                        7.4
Liczba_ocen                                                        1962
Liczba_czytelników                                                 4653
Name: 5419, dtype: object

In [ ]:
df.iloc[5416]

Link                  https://lubimyczytac.pl/oficjalne-recenzje-ksi...
Autor                                                         Shaun Tan
Tytuł                                              Zgubione, znalezione
Data_recenzji                                       2014-09-11 00:00:00
Treść_recenzji        Shaun Tan jest kapłanem wyobcowania. To znaczy...
Średnia_ocen                                                        8.2
Liczba_ocen                                                          91
Liczba_czytelników                                                  221
Name: 5420, dtype: object

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5968 entries, 0 to 5971
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Link                5968 non-null   object        
 1   Autor               5968 non-null   object        
 2   Tytuł               5968 non-null   object        
 3   Data_recenzji       5968 non-null   datetime64[ns]
 4   Treść_recenzji      5968 non-null   object        
 5   Średnia_ocen        5968 non-null   float64       
 6   Liczba_ocen         5968 non-null   int32         
 7   Liczba_czytelników  5968 non-null   int32         
dtypes: datetime64[ns](1), float64(1), int32(2), object(4)
memory usage: 373.0+ KB


In [ ]:
# Zmieniam typ zmiennej "Data_recenzji" podczas powyższych zmian wskoczył mi dodatkowo format godziny, nie potrzebuję tego.
df["Data_recenzji"] = pd.to_datetime(df["Data_recenzji"], format="%Y-%m-%d")

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5968 entries, 0 to 5971
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Link                5968 non-null   object        
 1   Autor               5968 non-null   object        
 2   Tytuł               5968 non-null   object        
 3   Data_recenzji       5968 non-null   datetime64[ns]
 4   Treść_recenzji      5968 non-null   object        
 5   Średnia_ocen        5968 non-null   float64       
 6   Liczba_ocen         5968 non-null   int32         
 7   Liczba_czytelników  5968 non-null   int32         
dtypes: datetime64[ns](1), float64(1), int32(2), object(4)
memory usage: 373.0+ KB


In [ ]:
# Sprawdzę czy dana książka ma tylko jedną recenzję czy więcej i zapisze te rekordy do osobnej df
# keep=False oznacza, że wszystkie wystąpienia duplikatów zostaną oznaczone jako True
df_duplicate_books = df[df.duplicated(subset=["Autor", "Tytuł"], keep=False)]

In [ ]:
# Nie ma tych rekordów dużo, więc chcę je wszystkie wyświetlić.
pd.set_option("display.max_rows", None)

In [ ]:
df_duplicate_books.sort_values(by=["Autor", "Tytuł", "Data_recenzji"])

Link  \
4650  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
325                https://lubimyczytac.pl/zanim-zgasna   
5868  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
2337  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
1285  https://lubimyczytac.pl/dzieje-domu-towarowego...   
1116  https://lubimyczytac.pl/dzieje-domu-towarowego...   
1231              https://lubimyczytac.pl/niezwyciezeni   
1173   https://lubimyczytac.pl/niezwyciezeni-2021-01-13   
4803  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
3708  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
5843  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
160   https://lubimyczytac.pl/otuchy-na-prozno-szukacie   
4562  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
1275  https://lubimyczytac.pl/swiadomosc-jako-blad-e...   
4826  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
1521  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
3715  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
2453  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
3984  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
994   https://lubimyczytac.pl/reczniki-w-dlon-kosmic...   
3137  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
642   https://lubimyczytac.pl/pragniemy-tylko-tego-c...   
4916  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
881          https://lubimyczytac.pl/blizny-przeszlosci   
3016  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
1137             https://lubimyczytac.pl/poza-schematem   
5066  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
1742  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
2583  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
306   https://lubimyczytac.pl/symbol-ludzkiej-egzyst...   
5823  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
3774  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
5381  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
1253   https://lubimyczytac.pl/kto-poslubi-panne-bennet   
5132  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
956    https://lubimyczytac.pl/nie-taki-diabel-straszny   
4426  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
2544  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
3331  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
2892  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
4296  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
1084    https://lubimyczytac.pl/koniec-lwowskiej-idylli   
4739  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
1379  https://lubimyczytac.pl/dziewczyna-utkana-z-pa...   
5842  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
2990  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
4289  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
869   https://lubimyczytac.pl/zycie-w-warszawskim-ge...   
3271  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
289                https://lubimyczytac.pl/czas-zbiorow   
4658  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
1291  https://lubimyczytac.pl/pod-maska-blazna-skryw...   
139   https://lubimyczytac.pl/z-wrogiem-pod-jedna-ba...   
87    https://lubimyczytac.pl/z-wrogiem-pod-jedna-ba...   
5621  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
2133  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
3644  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
724   https://lubimyczytac.pl/dziki-zachod-krzywd-i-...   
5207  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
3348  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
5829  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
2239  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
2846  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
669   https://lubimyczytac.pl/monachium-historyczne-...   
3796  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
505   https://lubimyczytac.pl/ludzka-pamiec-jest-zaw...   
4858  https://lubimyczytac.pl/oficjalne-recenzje-ksi...   
543   https://lubimyczytac.pl/historia-

In [ ]:
## ** opcjonalnie** Zapisuję do ramki danych i pliku xlsx te duplikaty
duplicate_books = pd.DataFrame(df_duplicate_books)
duplicate_books.to_excel("duplicate_books.xlsx")


### Jak widać, kilka tytułów (książek) ma więcej niż jedną tzw. oficjalna recenzję.
### Do tego w przypadku 6 książek, recenzja jest ta sama (ten sam autor i ta sama treść). Poza linkiem różnią się datą dodania.
### Muszę zwrócić uwagę, co do treści recenzji - sama treść jest identyczna, natomiast podczas scrapowania zaciągnął się fragment dotyczący ile osób poleca recenzje pod danym linkiem. I tu się rozjeżdża ta identyczność. Dlatego podczas usuwania rekordów nie będę odwoływała się do zmiennej 'Treść_recenzji'.
### Na potrzeby tego projektu, zostawię recenzje wcześniejsze (dotyczy to 6 rekordów).
### Pozostałe tytuły będą miały wszystkie zeskarpowane recenzje (mają róznych autorów, więc uważam, że warto to zostawić).

In [ ]:
# W związku z tym, że podczas scrapowania nie pobrałam autorów recenzji
#(żeby nie marnować czasu na ponowne scrapowanie - sprawdziłam te 6 tytułów w necie),
# do usunięcia zbędnych rekordów posłużę się zmiennymi "Autor", "Tytuł", "Data_recenzji" oraz indeksami.
rows_to_drop = df.query("index in [1116, 1173, 2892, 1379, 87, 3069]")


In [ ]:
# Usuwam rekordy dla indeksów: 1116, 1173, 2892, 1379, 87, 3069
to_drop = [1116, 1173, 2892, 1379, 87, 3069]
df = df.drop(to_drop, axis=0)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5962 entries, 0 to 5971
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Link                5962 non-null   object        
 1   Autor               5962 non-null   object        
 2   Tytuł               5962 non-null   object        
 3   Data_recenzji       5962 non-null   datetime64[ns]
 4   Treść_recenzji      5962 non-null   object        
 5   Średnia_ocen        5962 non-null   float64       
 6   Liczba_ocen         5962 non-null   int32         
 7   Liczba_czytelników  5962 non-null   int32         
dtypes: datetime64[ns](1), float64(1), int32(2), object(4)
memory usage: 372.6+ KB


# Dodam jeszcze zmienną z samym rokiem. Później może się przydać.

In [ ]:
df["Rok_recenzji"] = df["Data_recenzji"].dt.year

In [ ]:
# To będzie moja wyjściowa ramka danych do dalszych prac. Zapiszę ją do pliku csv.
reviews_eda1 = pd.DataFrame(df)
reviews_eda1.to_csv("reviews_eda1.csv")